### ISSUE TITLE
Refactor ServantError

### ISSUE THREAD

**Issue by dmjio** [haskell-servant/servant#158]
So for testing I'm using the servant-client generated functions w/ hspec. The issue is that hspec's `shouldBe` combinator requires an `Eq` instance for `ServantError`.
Here's an example
```haskell
------------------------------------------------------------------------------                                                                                                    
-- | Proxification                                                                                                                                                                
plusoneApi :: Proxy PlusOneAPI                                                                                                                                                    
plusoneApi = Proxy                                                                                                                                                                
------------------------------------------------------------------------------                                                                                                    
-- | Sigs                                                                                                                                                                         
plusOneDELETE :: Maybe JWT.JSON -> EitherT ServantError IO ()                                                                                                                     
plusOnePOST   :: PlusOneId      -> Maybe JWT.JSON -> EitherT ServantError IO PlusOne                                                                                              
plusOneGET    :: Maybe JWT.JSON -> EitherT ServantError IO PlusOne                                                                                                                
------------------------------------------------------------------------------                                                                                                    
-- | Client Handlers                                                                                                                                                              
plusOneGET :<|> plusOneDELETE :<|> plusOnePOST = client plusoneApi host                                                                                                           
  where host = BaseUrl Http "localhost" 8000                                                                                                                                      
------------------------------------------------------------------------------                                                                                                    
-- | Plus One Tests                                                                                                                                                               
plusOneWebTests :: Config -> SpecWith ()                                                                                                                                          
plusOneWebTests config = do                                                                                                                                                       
  describe "Plus One Web Tests" $ do                                                                                                                                              
    (uid, poid) <- runIO $ (,) <$> newUserId <*> newPlusOneId                                                                                                                     
    Right token <- runIO $ runApp config $ createToken uid                                                                                                                        
    it "Should create a plus one" $ do                                                                                                                                            
      result <- runEitherT $ plusOnePOST poid (Just token)                                                                                                                        
      result `shouldSatisfy` isRight                                                                                                                                              
    it "Should get a plus one" $ do                                                                                                                                               
      result <- runEitherT $ plusOneGET (Just token)                                                                                                                              
      result `shouldBe` Right (PlusOne uid poid)         -- <---- HERE                                                                                                                       
    it "Should delete a plus one" $ do                                                                                                                                            
      r1 <- runEitherT $ plusOneDELETE (Just token)                                                                                                                               
      r1 `shouldSatisfy` isRight                                                                                                                                                  
      r2 <- runEitherT $ plusOneGET (Just token)                                                                                                                                  
      r2 `shouldSatisfy` isLeft                                                                                                                                                                                         
```
And the error
```haskell
 No instance for (Eq ServantError) arising from a use of ‘shouldBe’ …                                                                  
    In a stmt of a 'do' block:                                                                                                                                                    
      result `shouldBe` Right (PlusOne uid poid)                                                                                                                                  
    In the second argument of ‘($)’, namely                                                                                                                                       
      ‘do { result <- runEitherT $ plusOneGET (Just token);                                                                                                                       
            result `shouldBe` Right (PlusOne uid poid) }’                                                                                                                         
    In a stmt of a 'do' block:                                                                                                                                                    
      it "Should get a plus one"                                                                                                                                                  
      $ do { result <- runEitherT $ plusOneGET (Just token);                                                                                                                      
             result `shouldBe` Right (PlusOne uid poid) }                                                                                                                         
Compilation failed.                                  
```
The problem is that `ServantError` has `HttpException` on the connection failure type. What if we always just returned False for that case? Since exceptions really can't be compared until cast.

If you agree I can submit a pull.

**Comment by alpmestan** [comment]
I don't mind a sensible `Eq` instance personally.

**Comment by jkarni** [comment]
:+1: - I've wanted this myself during tests.

**Comment by dmjio** [comment]
can this be closed?

**Comment by alpmestan** [comment]
We still don't have an `Eq` instance, as evidenced [here](https://github.com/haskell-servant/servant/blob/master/servant-client/src/Servant/Common/Req.hs#L56).

**Comment by jkarni** [comment]
That datatype is pretty ugly (records on sum type?), and begging to be reconsidered. How about:

```
data ServantError = HttpError ServantErr | ConnectionError Text
```
And we just `show` the exception? 

Another thing to consider is to not have a separate datatype for ServantError at all, and just use `ServantErr`. We just wouldn't catch connection error exceptions.

**Comment by alpmestan** [comment]
We have to catch them "unfortunately". I agree however that we really ought to make this bit of `servant-client` nicer.

**Comment by jkarni** [comment]
I'll rename this to `Refactor ServantError`?

**Comment by alpmestan** [comment]
Yes.

**Comment by jkarni** [comment]
@alpmestan I think we discussed this elsewhere, but I don't remember the details. Why do we have to catch connection errors?

**Comment by alpmestan** [comment]
IIRC, it was because if we don't, people have do handle two levels of errors, each in a different way. The first is the `Left` branch of the `Either`, the second is exceptions thrown by *http-client*. I personally would find it terrible, as a user.

**Comment by jkarni** [comment]
But those *are* very different errors conceptuall. For example, if I want to test that two servers behave the same way, I really just want equality over the Either *if* it does not include connection errors. If it does include connection errors, I have to manually check that there weren't any. Moreover, the logic that handles connection errors will invariably be different - one is "check the status code and see what you should do", the other is "retry" (modulo user-client errors).

**Comment by jkarni** [comment]
In short - my opinion is that what `http-client` does is correct, and our attempts to change it are wrong.

**Comment by jkarni** [comment]
(And there's also the very elegant fact that we could have handlers and client functions have the same type, fwiw.)

**Comment by alpmestan** [comment]
Go for it then.

**Comment by soenkehahn** [comment]
Yes, I think being able to have the client functions and the server handlers have exactly the same type is kind of a killer argument here. You could imagine some common code that is used on both sides that uses some operations that on the server side are provided by normal server handlers, but on the client side by `servant-client`. To allow that would be awesome.

Of course the common `ServantErr` type then has to include network errors. And then a server handler can theoretically throw a network error. We have to make sure that these network errors aren't propagated as a network error to `servant-client`, because that'd be lying. I would be fine with just have a runtime error in that case (`500` for the client).

**Comment by soenkehahn** [comment]
(Talking about this with @jkarni offline.)

I didn't consider the possibility to throw exceptions in case of network errors. That might also be acceptable. My main goal is to unify the error types for `servant-server` and `servant-client`.

**Comment by alpmestan** [comment]
I don't think we should be able to throw network error exceptions from server handlers, but I otherwise agree.

**Comment by jkarni** [comment]
> I don't think we should be able to throw network error exceptions from server handlers, but I otherwise agree.

We can't really prevent that :)

**Comment by alpmestan** [comment]
By "throw", I meant return `Left someNetworkError`.

**Comment by jkarni** [comment]
Oh. Agreed.

**Comment by fizruk** [comment]
Below are some of my thoughts/fantasies, based on APIs I have to deal with in real world.

Some (most?) APIs tend to return 200 status code even when there has been an error. Typically I see this as `{"error": "no credit"}` for failure and `{"result": 123}` for success. This is probably just bad design, but it happens and I have to account for it when integrating with such APIs.

So since you're going to redesign the `ServantError`, I would like to discuss a solution to this problem as well. One way to handle this is, of course, to simply declare an `Either`-like type with custom `ToJSON`/`FromJSON` instances, but I can think of some problems with that approach:
- the response body is an error, so it should be `Left`, not `Right` (in `servant-client`);
- these error should be easily "thrown" in server handlers just like other errors, not `return`ed (this point may be omitted since returning errors with 200 status might not be a good idea, so we might not want to support it in `servant-server`).

We could probably declare an `{-# OVERLAPPING #-}` `HasClient` instance for when that `Either`-like type appears in the response type to convert status 200 failures into `Left` in `servant-client`, but there still has to be some room for these user errors. Currently I think 200 errors can be placed under `FailureResponse`.

Irrelevant of the 200 errors, one thing I don't like about `FailureResponse` is that `responseBody` is any `ByteString`. For APIs I work with *I know* which errors might occur most of the time, so it appears we can store errors in a more meaningful way.
I think that maybe we could use `ServantError` as a "low-level" error representation (like `Value` for JSON values). Then we might be able to decode (some) errors and get better client functions:

```haskell
-- | Somewhere in servant.
class FromServantError a where
  -- | Try to convert a ServantError into a more sensible thing.
  parseServantError :: ServantError -> Maybe a

-- | An API for sending a text message to a phone number.
type SendAPI
  = QueryParam "phone" Phone
  :> QueryParam "text" Text
  :> Post '[JSON] (MessageId `WithError` SendError)

-- | Business-logic-related errors (parameters validation is not handled here).
data SendError
  = NoCredit         -- no credit on the account to send a message
  | AccountSuspended -- unable to send messages from this account due to a suspension

-- some 503 errors might get decoded into SendError
instance FromServantError SendError where ...

-- send :: Client SendAPI
send :: Maybe Phone  -- ^ phone number
     -> Maybe Text   -- ^ message to send
     -> EitherT (ServantError, Maybe SendError) IO MessageId
send = ...
```

The paragraph above may be hard to digest, because I myself struggle to formulate what I really want.
Nonetheless, I believe we can handle business-logic errors better than we do currently!

**Comment by alpmestan** [comment]
@fizruk I have dealt with many such APIs in the past as well. I do not think the need you're expressing is far fetched and finding a nice solution for this would be great.

**Comment by fizruk** [comment]
@alpmestan should this problem be discussed here, or in a separate issue?

**Comment by alpmestan** [comment]
I'm not sure. Depends on whether @jkarni thinks this belongs here, I guess ?

**Comment by jkarni** [comment]
I'm in favour of separating these issues (partly out of self-interest - I now need an `Eq` instance for `ServantError` before releasing `servant-comock` (or more likely `servant-quickcheck`)). 

@haskell-servant/maintainers Have we decided what to do here? The options I see are:

(1) Keep the datatype as is, but make the `ConnectionError` case testable for equality in some way (e.g. by `show`-ing it).
(2) Remove the `ConnectionError` constructor, and make that be an actual exception.
(3) Remove everything except `FailureReponse`, and make all of them exceptions.

(Independently of that, the non-lens partial record accessors should be fixed/removed, in my opinion)

**Comment by soenkehahn** [comment]
I'd be fine with merging (1) and (2) in separate PRs independently of each other.

I don't see (3) as an improvement, but I also don't know, why you're proposing it. Do you want to get rid of the partial selectors? If so, I think you should open another issue/PR and discuss there. Same for unifying the two error types.

(I think this issue has gotten too scope creepy. I'd vote for reverting this to being about having `Eq ServantError` and nothing else. Don't get me wrong, I think the above discussions (and others that are similar) are useful. But they can stand in the way of focusing on solving the problem at hand. Not sure how to deal with this dilemma in general...)

**Comment by jkarni** [comment]
> I think this issue has gotten too scope creepy. I'd vote for reverting this to being about having Eq ServantError and nothing else.

Aren't (1)-(3) the options for doing that?

**Comment by soenkehahn** [comment]
@jkarni: Sorry, I read (2) as not adding a an `Eq` instance, but just removing `ConnectionError` and throwing an exception in that case. But yes, once that constructor is removed, `Eq` should be derivable.

Regarding (3): The problem for `Eq` is `SomeException`, right? So removing `ConnectionError` is all we need.

**Comment by soenkehahn** [comment]
I've created #481 implementing (1).

**Comment by soenkehahn** [comment]
Fixed in #481. (Please open other issues/PRs for further changes to `ServantError`.)
